### PR TITLE
KYLIN-4829 Support to use thread-level SparkSession to execute query

### PR DESCRIPTION
**Pull request by zzcclp**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [x] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zzcclp** [comment]
Ran test cases local successfully.

**Comment by zzcclp** [comment]
## The Results of Testing Manually

### Test Env

- Hadoop 2.7.0 on docker.
- Commit : [3b3786c5c](https://github.com/apache/kylin/commit/3b3786c5c9602838cd4abd0a6d40574550ec8622)
- Sparder Env : 
   spark.executor.cores=1
   spark.executor.instances=4
   spark.executor.memory=2G
   spark.executor.memoryOverhead=1G
   spark.sql.shuffle.partitions=4


### Before this patch
The shuffle partition number of all querys is 4, which equals to the total cores number.
image

image

image

image


### After this patch
The shuffle partition number of each query is calculated according to the scanned bytes of each query:
image

image

image

image

**Comment by zzcclp** [comment]
## Performance Testing

### Test Env

- Hadoop 2.7.0 on docker.
- Commit : [ed0649b1](https://github.com/apache/kylin/pull/1495/commits/ed0649b140529bdfafea8cce846962b6ca9c3f73)
- Sparder Env : 
   spark.executor.cores=1
   spark.executor.instances=6
   spark.executor.memory=2G
   spark.executor.memoryOverhead=1G
   spark.sql.shuffle.partitions=6

### Test
Sends 5 SQLs at the same time

### Before this patch
The shuffle partition number of all querys is 6, which equals to the total cores number, and totally spent 3.09s to finish these 5 SQLs.

5 SQLs submitted at the same time.
image

image

SQL1, there are 6 partitions at the second stage, even though there are only 216KB data to read.
image


SQL2, there are still 6 partitions at the second stage, even though there are only 262KB data to read.
image


### After this patch
The shuffle partition number of each query is calculated according to the scanned bytes of each query, and totally spent 2.55s to finish these 5 SQLs.

5 SQLs submitted at the same time.
image

image


SQL1, there is only 1 partition at the second stage.
image


SQL2, there is only 1 partition at the second stage.
image

File: kylin-spark-project/kylin-spark-query/src/test/java/org/apache/spark/sql/SparderContextFacadeTest.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            if (service.take().get() != null) {
+                hasError = true;
+            }
+        }
+        Assert.assertTrue(hasError);
+
+        executor.shutdown();
+    }
+
+    protected CompletionService<Throwable> runThreadSparkSessionTest(ThreadPoolExecutor executor,
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-query/src/test/java/org/apache/spark/sql/SparderContextFacadeTest.java:92]: Looks like `isOriginal` is always `false` ?

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/spark/sql/KylinSparkEnv.scala [hunk 2]
```diff
@@ -25,96 +25,27 @@ object KylinSparkEnv extends Logging {
 	@volatile
 	private var spark: SparkSession = _
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/spark/sql/KylinSparkEnv.scala:26]: Are you plan to delete `KylinSparkEnv` ?

File: kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala [hunk 3]
```diff
@@ -17,25 +17,26 @@
...
 import org.apache.spark.sql.SparkSession
+import org.apache.spark.utils.SparderUtils
 
 trait ResetShufflePartition extends Logging {
+  val PARTITION_SPLIT_BYTES: Long = KylinConfig.getInstanceFromEnv.getQueryPartitionSplitSizeMB * 1024 * 1024 // 64MB
 
   def setShufflePartitions(bytes: Long, sparkSession: SparkSession): Unit = {
     QueryContextFacade.current().addAndGetSourceScanBytes(bytes)
-    val defaultParallelism = sparkSession.sparkContext.defaultParallelism
+    val defaultParallelism = SparderUtils.getTotalCore(sparkSession.sparkContext.getConf)
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala:30]: Why change from  `defaultParallelism` to `getTotalCore` ?

File: kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala [hunk 4]
```diff
@@ -17,25 +17,26 @@
...
     val kylinConfig = KylinConfig.getInstanceFromEnv
     val partitionsNum = if (kylinConfig.getSparkSqlShufflePartitions != -1) {
       kylinConfig.getSparkSqlShufflePartitions
     } else {
-      Math.min(QueryContextFacade.current().getSourceScanBytes / (
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala:33]: Looks like the original code is wrong, if `sparkContext.defaultParallelism` is 1, `partitionsNum` will always be 1.

File: kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala [hunk 5]
```diff
@@ -17,25 +17,26 @@
...
-        KylinConfig.getInstanceFromEnv.getQueryPartitionSplitSizeMB * 1024 * 1024 * 2) + 1,
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala:34]: You remove `* 2`, is it correct ?

**Review by hit-lacus — APPROVED** [review]
It looks almost good to me. Thank you @zzcclp .

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/spark/sql/KylinSparkEnv.scala (hunk 2 above)
Comment (zzcclp) [reply kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/spark/sql/KylinSparkEnv.scala:26]: Yeah, will raise another pr to do this, it needs to refactor some modules

File: kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala (hunk 3 above)
Comment (zzcclp) [reply kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala:30]: Make it the same algorithm as the code of 'ResultPlan.collectInternal'.

File: kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala (hunk 4 above)
Comment (zzcclp) [reply kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala:33]: Using 'Math.min' is right, it makes sure that the max partition number is not larger than the total cores.

File: kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala (hunk 5 above)
Comment (zzcclp) [reply kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/ResetShufflePartition.scala:34]: '* 2' doesn't make sense. If users want to use larger partition split size, they can increase the value of 'kylin.query.spark-engine.partition-split-size-mb' .

**Pull request merged**